src/components/TypingText/TypingText.tsx

```diff
@@ -1,52 +1,238 @@
-  };
+const normalizeMarkdown = (rawText: string): string => {
+  return rawText
+    .trim()
+    .replace(/\r\n|\r/g, '\n')
+    .split('\n')
+    .map((line) => line.trim())
+    .join('\n')
+    .replace(/\n\s*\n\s*\n/g, '\n\n');
+};
+
+/**
+ * TypingText Component
+ * Renders markdown text with a typewriter effect
+ *
+ * @param {Object} props - The properties object
+ * @param {string} props.text - The initial markdown text to be displayed
+ * @param {string} props.id - Unique identifier for the text
+ * @param {number} [props.speed=TYPING_SPEEDS.MIN_SPEED] - Typing speed in milliseconds
+ * @param {boolean} props.isUserMessageSent - Flag indicating if a (new) user message is sent
+ * @param {function} props.onTypingStart - Callback function to be called when typing starts
+ * @param {number} [props.chatIndex=1] - Index of the chat
+ * @param {function} props.onTypingComplete - Callback function to be called when typing completes
+ * @returns {JSX.Element} The TypingText component
+ */
+export default function TypingText({
+  text: initialText,
+  id,
+  speed = TYPING_SPEEDS.MIN_SPEED,
+  isUserMessageSent,
+  onTypingStart,
+  chatIndex = 1,
+  onTypingComplete,
+}: TypingTextProps) {
+  const [displayedText, setDisplayedText] = useState('');
+  const normalizedText = normalizeMarkdown(initialText);
 
   useEffect(() => {
-    let index = 0;
-
-    const typeCharacter = (): void => {
-      if (index < text.length - 1) {
-        const currentChar = text[index];
-        setDisplayedText((prev) => prev + text[index]);
-        index += 1;
-        setTimeout(typeCharacter, getCharacterSpeed(currentChar)); // Generate a new timeout with random speed
-      } else if (onTypingComplete) {
-        onTypingComplete(); // Call onTypingComplete callback when typing is complete
+    let currentIndex = 0;
+
+    const completionKey = STORAGE_KEYS.COMPLETION(id);
+    const progressKey = STORAGE_KEYS.PROGRESS(id);
+
+    /**
+     * Runs at the beginning of typing animation
+     * sets isTyping to false so that dot animation
+     * gets cancelled
+     */
+    const handleTypingStart = async () => {
+      if (onTypingStart) {
+        await onTypingStart(chatIndex, false);
       }
     };
 
-    typeCharacter(); // Start typing
+    /**
+     * Calculates typing delay based on progress for
+     * dynamic typing animation
+     */
+    const calculateDelay = (progress: number): number => {
+      const progressRatio = progress / normalizedText.length;
+      return speed + progressRatio * progressRatio * (TYPING_SPEEDS.MAX_SPEED - TYPING_SPEEDS.MIN_SPEED);
+    };
 
+    /**
+     * Handles typing animation completion
+     */
+    const handleTypingComplete = () => {
+      localStorage.setItem(completionKey, 'true');
+      localStorage.removeItem(progressKey);
+      onTypingComplete?.();
+    };
+
+    /**
+     * Restores typing progress from localStorage
+     */
+    const restoreProgress = () => {
```

Comment:
return type of the function + this function could be extracted into operations file